Proposal: Reduce calls to pkg management

I've always noticed how many times the bootstrap called the package management
For another PR I'm now looking at the Fedora piece, and notice that it's quite easy to prevent most of those.

Using the example below I've downed the number of 6 dnf sessions back to two with no meaningful change in end-result/failure modes AFAICT.

So this is just cooked up, but if I made this into a decent PR, would that be acceptable? 

### EXAMPLE (WIP/UNTESTED)
```
#   To Install Dependencies, which is required, one of:
#       1. install_<distro>_<major_version>_<install_type>_deps
#       2. install_<distro>_<major_version>_<minor_version>_<install_type>_deps
#       3. install_<distro>_<major_version>_deps
#       4  install_<distro>_<major_version>_<minor_version>_deps
#       5. install_<distro>_<install_type>_deps
#       6. install_<distro>_deps
#
#   To install salt, which, of course, is required, one of:
#       1. install_<distro>_<major_version>_<install_type>
#       2. install_<distro>_<major_version>_<minor_version>_<install_type>
#       3. install_<distro>_<install_type>

```
For a typical git install on Fedora,
`install_git_fedora_deps()`
`install_fedora_deps()`
etc
#### NOW
```sh
install_fedora_git_deps() {

    if [ "$_INSECURE_DL" -eq $BS_FALSE ] && [ "${_SALT_REPO_URL%%://*}" = "https" ]; then
        dnf install -y ca-certificates || return 1
    fi

    install_fedora_deps || return 1

    if ! __check_command_exists git; then
        dnf install -y git || return 1
    fi

    __git_clone_and_checkout || return 1

    __PACKAGES="python2-tornado systemd-python"

    if [ "$_INSTALL_CLOUD" -eq $BS_TRUE ]; then
        __PACKAGES="${__PACKAGES} python-libcloud python-netaddr"
    fi

    # shellcheck disable=SC2086
    dnf install -y ${__PACKAGES} || return 1

    # Let's trigger config_salt()
    if [ "$_TEMP_CONFIG_DIR" = "null" ]; then
        _TEMP_CONFIG_DIR="${_SALT_GIT_CHECKOUT_DIR}/conf/"
        CONFIG_SALT_FUNC="config_salt"
    fi

    return 0
}


install_fedora_deps() {

    __PACKAGES="dnf-utils libyaml m2crypto PyYAML python-crypto python-jinja2"
    __PACKAGES="${__PACKAGES} python2-msgpack python2-requests python-zmq"

    # shellcheck disable=SC2086
    dnf install -y ${__PACKAGES} || return 1

    if [ "$_UPGRADE_SYS" -eq $BS_TRUE ]; then
        dnf -y update || return 1
    fi

    if [ "${_EXTRA_PACKAGES}" != "" ]; then
        echoinfo "Installing the following extra packages as requested: ${_EXTRA_PACKAGES}"
        # shellcheck disable=SC2086
        dnf install -y ${_EXTRA_PACKAGES} || return 1
    fi

    return 0
}
```

### THEN
```sh
install_fedora_git_deps() {

    __PACKAGES="${__PACKAGES} python2-tornado systemd-python"
    if [ "$_INSECURE_DL" -eq $BS_FALSE ] && [ "${_SALT_REPO_URL%%://*}" = "https" ]; then
        __PACKAGES="${__PACKAGES} ca-certificates"
    fi
    if [ "$_INSTALL_CLOUD" -eq $BS_TRUE ]; then
        __PACKAGES="${__PACKAGES} python-libcloud python-netaddr"
    fi
    if ! __check_command_exists git; then
        __PACKAGES="${__PACKAGES} git"
    fi
    install_fedora_deps || return 1

    __git_clone_and_checkout || return 1

    # Let's trigger config_salt()
    if [ "$_TEMP_CONFIG_DIR" = "null" ]; then
        _TEMP_CONFIG_DIR="${_SALT_GIT_CHECKOUT_DIR}/conf/"
        CONFIG_SALT_FUNC="config_salt"
    fi

    return 0
}

install_fedora_deps() {
    if [ "$_UPGRADE_SYS" -eq $BS_TRUE ]; then
        dnf -y update || return 1
    fi

    __PACKAGES="${__PACKAGES} dnf-utils libyaml m2crypto PyYAML python-crypto python-jinja2"
    __PACKAGES="${__PACKAGES} python2-msgpack python2-requests python-zmq"

    if [ "${_EXTRA_PACKAGES}" != "" ]; then
        echoinfo "Installing the following extra packages as requested: ${_EXTRA_PACKAGES}"
        __PACKAGES="${__PACKAGES} ${_EXTRA_PACKAGES}"
    fi
    # shellcheck disable=SC2086
    dnf install -y ${__PACKAGES} || return 1

    return 0
}
```
## RESULT
Instead of 6 dnf sessions we now have 2, with AFAICT the exact same result.

--- a/bootstrap-salt.sh
+++ b/bootstrap-salt.sh
@@ -3415,35 +3415,31 @@ install_debian_check_services() {
 #
 
 install_fedora_deps() {
+    if [ "$_UPGRADE_SYS" -eq $BS_TRUE ]; then
+        dnf -y update || return 1
+    fi
+
     if [ -n "$_PY_EXE" ] && [ "$_PY_MAJOR_VERSION" -eq 3 ]; then
         # Packages are named python3-<whatever>
         PY_PKG_VER=3
-        __PACKAGES="python3-m2crypto python3-PyYAML"
+        __PACKAGES="${__PACKAGES} python3-m2crypto python3-PyYAML"
     else
         PY_PKG_VER=2
-        __PACKAGES="m2crypto"
+        __PACKAGES="${__PACKAGES} m2crypto"
         if [ "$DISTRO_MAJOR_VERSION" -ge 28 ]; then
           __PACKAGES="${__PACKAGES} python2-pyyaml"
         else
           __PACKAGES="${__PACKAGES} PyYAML"
        fi
     fi
-
-    __PACKAGES="${__PACKAGES} procps-ng dnf-utils libyaml python${PY_PKG_VER}-crypto python${PY_PKG_VER}-jinja2"
+    __PACKAGES="${__PACKAGES} dnf-utils libyaml procps-ng python${PY_PKG_VER}-crypto python${PY_PKG_VER}-jinja2"
     __PACKAGES="${__PACKAGES} python${PY_PKG_VER}-msgpack python${PY_PKG_VER}-requests python${PY_PKG_VER}-zmq"
-
-    # shellcheck disable=SC2086
-    dnf install -y ${__PACKAGES} || return 1
-
-    if [ "$_UPGRADE_SYS" -eq $BS_TRUE ]; then
-        dnf -y update || return 1
-    fi
-
     if [ "${_EXTRA_PACKAGES}" != "" ]; then
         echoinfo "Installing the following extra packages as requested: ${_EXTRA_PACKAGES}"
-        # shellcheck disable=SC2086
-        dnf install -y ${_EXTRA_PACKAGES} || return 1
     fi
+
+    # shellcheck disable=SC2086
+    dnf install -y ${__PACKAGES} ${_EXTRA_PACKAGES} || return 1
 
     return 0
 }
@@ -3494,35 +3490,37 @@ install_fedora_git_deps() {
         PY_PKG_VER=2
     fi
 
+    __PACKAGES=
     if [ "$_INSECURE_DL" -eq $BS_FALSE ] && [ "${_SALT_REPO_URL%%://*}" = "https" ]; then
-        dnf install -y ca-certificates || return 1
+        __PACKAGES="${__PACKAGES} ca-certificates"
+    fi
+    if ! __check_command_exists git; then
+        __PACKAGES="${__PACKAGES} git"
+    fi
+    if [ "$_INSTALL_CLOUD" -eq $BS_TRUE ]; then
+        __PACKAGES="${__PACKAGES} python${PY_PKG_VER}-libcloud python${PY_PKG_VER}-netaddr"
+    fi
+    __PACKAGES="${__PACKAGES} python${PY_PKG_VER}-systemd"
+
+    # Fedora 28+ ships with tornado 5.0+ which is broken for salt on py3
+    # https://github.com/saltstack/salt-bootstrap/issues/1220
+    if [ "${PY_PKG_VER}" -lt 3 ] || [ "$DISTRO_MAJOR_VERSION" -lt 28 ]; then
+        __PACKAGES="${__PACKAGES} python${PY_PKG_VER}-tornado"
     fi
 
     install_fedora_deps || return 1
 
-    if ! __check_command_exists git; then
-        dnf install -y git || return 1
-    fi
-
     __git_clone_and_checkout || return 1
 
-    __PACKAGES="python${PY_PKG_VER}-systemd"
+    # Fedora 28+ needs tornado <5.0 from pip
+    # https://github.com/saltstack/salt-bootstrap/issues/1220
     if [ "${PY_PKG_VER}" -eq 3 ] && [ "$DISTRO_MAJOR_VERSION" -ge 28 ]; then
         __check_pip_allowed "You need to allow pip based installations (-P) for Tornado <5.0 in order to install Salt on Python 3"
         grep tornado "${_SALT_GIT_CHECKOUT_DIR}/requirements/base.txt" | while IFS='
 '         read -r dep; do
             "${_PY_EXE}" -m pip install "${dep}" || return 1
         done
-    else
-        __PACKAGES="${__PACKAGES} python${PY_PKG_VER}-tornado"
     fi
-
-    if [ "$_INSTALL_CLOUD" -eq $BS_TRUE ]; then
-        __PACKAGES="${__PACKAGES} python${PY_PKG_VER}-libcloud python${PY_PKG_VER}-netaddr"
-    fi
-
-    # shellcheck disable=SC2086
-    dnf install -y ${__PACKAGES} || return 1
 
     # Let's trigger config_salt()
     if [ "$_TEMP_CONFIG_DIR" = "null" ]; then

--- a/bootstrap-salt.sh
+++ b/bootstrap-salt.sh
@@ -3419,6 +3419,7 @@ install_fedora_deps() {
         dnf -y update || return 1
     fi
 
+    __PACKAGES="${__PACKAGES:=}"
     if [ -n "$_PY_EXE" ] && [ "$_PY_MAJOR_VERSION" -eq 3 ]; then
         # Packages are named python3-<whatever>
         PY_PKG_VER=3
@@ -3490,7 +3491,7 @@ install_fedora_git_deps() {
         PY_PKG_VER=2
     fi
 
-    __PACKAGES=
+    __PACKAGES="${__PACKAGES:=}"
     if [ "$_INSECURE_DL" -eq $BS_FALSE ] && [ "${_SALT_REPO_URL%%://*}" = "https" ]; then
         __PACKAGES="${__PACKAGES} ca-certificates"
     fi

